be/src/formats/orc/utils.h

```diff
@@ -26,13 +23,63 @@
 #include "formats/orc/orc_mapping.h"
 #include "formats/orc/utils.h"
 #include "gen_cpp/orc_proto.pb.h"
-#include "runtime/types.h"
+#include "io/shared_buffered_input_stream.h"
 #include "types/date_value.h"
-#include "types/logical_type.h"
 #include "types/timestamp_value.h"
 
 namespace starrocks {
 
+class DiskRange {
+public:
+    DiskRange(int64_t off, int64_t len) : offset(off), length(len) {
+        DCHECK(off > 0);
+        DCHECK(len > 0);
+    }
+
+    void operator=(const DiskRange& other) {
+        offset = other.offset;
+        length = other.length;
+    }
+
+    /**
+    * Returns the minimal DiskRange that encloses both this DiskRange
+    * and otherDiskRange. If there was a gap between the ranges the
+    * new range will cover that gap.
+    */
+    DiskRange span(const DiskRange& otherDiskRange) const {
+        const int64_t start = std::min(offset, otherDiskRange.offset);
+        const int64_t end = std::max(get_end(), otherDiskRange.get_end());
+        return DiskRange(start, end - start);
+    }
+
+    int64_t get_end() const { return offset + length; }
+
+    int64_t offset;
+    int64_t length;
+};
+
+class DiskRangeHelper {
+public:
+    static void mergeAdjacentDiskRanges(std::vector<io::SharedBufferedInputStream::IORange>& io_ranges,
+                                        const std::vector<DiskRange>& disk_ranges, const int64_t max_merge_distance,
+                                        const int64_t max_merged_size) {
+        if (disk_ranges.empty()) {
+            return;
+        }
+        DiskRange last = disk_ranges[0];
+        for (size_t i = 1; i < disk_ranges.size(); i++) {
+            DiskRange current = disk_ranges[i];
+            DiskRange merged = last.span(current);
+            if (merged.length <= max_merged_size && last.get_end() + max_merge_distance >= current.offset) {
+                last = merged;
+            } else {
+                io_ranges.emplace_back(last.offset, last.length, true);
+                last = current;
+            }
```

Comment:
The `last` range will be lost if all disk ranges merged into it and never emplace_back to io_ranges.

Reply:
fixed